pkg/controller/clusternamespacedeletion/manager.go

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Red Hat, Inc.
+// Copyright Contributors to the Open Cluster Management project
+
+package clusternamespacedeletion
+
+import (
+	asv1beta1 "github.com/openshift/assisted-service/api/v1beta1"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/stolostron/managedcluster-import-controller/pkg/helpers"
+	addonv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const controllerName = "clusternamespacedeletion-controller"
+
+// Add creates a new managedcluster controller and adds it to the Manager.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager, clientHolder *helpers.ClientHolder,
+	importSecretInformer, autoImportSecretInformer cache.SharedIndexInformer) (string, error) {
+	return controllerName, add(mgr, newReconciler(clientHolder))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(clientHolder *helpers.ClientHolder) reconcile.Reconciler {
+	return &ReconcileClusterNamespaceDeletion{
+		client:   clientHolder.RuntimeClient,
+		recorder: helpers.NewEventRecorder(clientHolder.KubeClient, controllerName),
+	}
+}
+
+// adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New(controllerName, mgr, controller.Options{
+		Reconciler:              r,
+		MaxConcurrentReconciles: helpers.GetMaxConcurrentReconciles(),
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &clusterv1.ManagedCluster{}},
+		handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Name: o.GetName(),
+					},
+				},
+			}
+		}),
+	); err != nil {
+		return err
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &corev1.Namespace{}},
```

Comment:
It seems we don't need to watch the ns,

Reply:
I think we need to in case controller restarted